evm/contracts/Coordinator.sol

```diff
@@ -120,45 +113,43 @@ contract Coordinator is ChainlinkRequestInterface, CoordinatorInterface {
    * @notice Stores a Service Agreement which has been signed by the given oracles
    * @dev Validates that each oracle has a valid signature.
    * Emits NewServiceAgreement event.
-   * @param _payment The amount of payment given (specified in wei)
-   * @param _expiration The expiration that nodes should respond by
-   * @param _endAt The date which the service agreement is no longer valid
-   * @param _oracles Array of oracle addresses which agreed to the service agreement
-   * @param _vs Array of recovery IDs of the oracle signatures
-   * @param _rs Array of first 32 bytes of the oracle signatures
-   * @param _ss Array of second 32 bytes of the oracle signatures
-   * @param _requestDigest Hash of the normalized job specification
+   * @param _agreement The Service Agreement to be initiated
+   * @param _signatures The signatures of the oracles in the agreement
    * @return The Service Agreement ID
    */
   function initiateServiceAgreement(
-    uint256 _payment,
-    uint256 _expiration,
-    uint256 _endAt,
-    address[] _oracles,
-    uint8[] _vs,
-    bytes32[] _rs,
-    bytes32[] _ss,
-    bytes32 _requestDigest
+    ServiceAgreement memory _agreement,
+    OracleSignatures memory _signatures
   )
-    external
+    public 
     returns (bytes32 serviceAgreementID)
   {
-    require(_oracles.length == _vs.length && _vs.length == _rs.length && _rs.length == _ss.length, "Must pass in as many signatures as oracles"); /* solium-disable-line max-len */
-    require(_endAt > block.timestamp, "End of ServiceAgreement must be in the future");
-
-    serviceAgreementID = getId(_payment, _expiration, _endAt, _oracles, _requestDigest);
-
-    registerOracleSignatures(serviceAgreementID, _oracles, _vs, _rs, _ss);
+    require(
+      _agreement.oracles.length == _signatures.vs.length && 
+      _signatures.vs.length == _signatures.rs.length && 
+      _signatures.rs.length == _signatures.ss.length, 
+      "Must pass in as many signatures as oracles"
+    ); 
+    require(_agreement.endAt > block.timestamp, "End of ServiceAgreement must be in the future");
+
+    serviceAgreementID = getId(
```

Comment:
I think it'll actually save us gas, especially in the long term as we add parameters to the struct, since struct params are tightly-packed. I made [this repo](https://github.com/thodges-gh/struct-params) to see gas usage for passing in params vs a struct to a function.

Reply:
I need to read the bytecode that is generated. I'm guessing that there is some extra serialization overhead, but I'm guessing that using structs will improve Solidity's stack usage throughout the external function call.

Reply:
@jalextowle can you look into which is cheaper? It would be convenient to pass structs around, but if it's more expensive then we have a good reason not to.